src/model_tool.py

```diff
@@ -225,3 +225,63 @@ def score_auc(self, texts, labels):
+        # context vector
+        attention_mul = Multiply()([input_tensor, attention_probs])
+        return {'attention_probs': attention_probs, 'attention_preds': attention_mul}
+
+    def build_probs(self):
+        sequence_input = Input(shape=(self.hparams['max_sequence_length'],), dtype='int32')
+        embedding_layer = Embedding(len(self.tokenizer.word_index) + 1,
+                                    self.hparams['embedding_dim'],
+                                    weights=[self.embedding_matrix],
+                                    input_length=self.hparams['max_sequence_length'],
+                                    trainable=self.hparams['embedding_trainable'])
+
+        embedded_sequences = embedding_layer(sequence_input)
+        x = embedded_sequences
+        for filter_size, kernel_size, pool_size in zip(self.hparams['cnn_filter_sizes'], self.hparams['cnn_kernel_sizes'], self.hparams['cnn_pooling_sizes']):
+            x = self.build_conv_layer(x, filter_size, kernel_size, pool_size)
+
+        x = Flatten()(x)
+        x = Dropout(self.hparams['dropout_rate'], name="Dropout")(x)
+        # TODO(nthain): Parametrize the number and size of fully connected layers
+        x = Dense(250, activation='relu', name="Dense_RELU")(x)
+
+        attention_dict = self.build_dense_attention_layer(x)
+        preds = attention_dict['attention_probs']
+        preds = Dense(2, name="preds_dense")(preds)
+        rmsprop = RMSprop(lr=self.hparams['learning_rate'])
+        self.model = Model(sequence_input, preds)
+        self.model.compile(loss='categorical_crossentropy', optimizer=rmsprop,metrics=['acc'])
+
+    def build_model(self):
+        print('print inside build model')
+        sequence_input = Input(shape=(self.hparams['max_sequence_length'],), dtype='int32')
+        embedding_layer = Embedding(len(self.tokenizer.word_index) + 1,
+                                    self.hparams['embedding_dim'],
+                                    weights=[self.embedding_matrix],
+                                    input_length=self.hparams['max_sequence_length'],
+                                    trainable=self.hparams['embedding_trainable'])
+
+        embedded_sequences = embedding_layer(sequence_input)
+        x = embedded_sequences
+        for filter_size, kernel_size, pool_size in zip(self.hparams['cnn_filter_sizes'], self.hparams['cnn_kernel_sizes'], self.hparams['cnn_pooling_sizes']):
+            x = self.build_conv_layer(x, filter_size, kernel_size, pool_size)
+
+        x = Flatten()(x)
+        x = Dropout(self.hparams['dropout_rate'], name="Dropout")(x)
+        x = Dense(250, activation='relu', name="Dense_RELU")(x)
+
+        attention_dict = self.build_dense_attention_layer(x)
+        preds = attention_dict['attention_preds']
+        preds = Dense(2, name="preds_dense", activation='softmax')(preds)
+        rmsprop = RMSprop(lr=self.hparams['learning_rate'])
+        self.model = Model(sequence_input, preds)
```

Comment:
It would be nice to expose the attention weights as well as the predictions so we can visualize what the model is paying attention to.

Reply:
I now save this [here](https://github.com/rubinovitz/unintended-ml-bias-analysis/blob/a61a681d5f16bb544e58b2c07e7e336a804e942b/unintended_ml_bias/model_tool.py#L465).